File: Generative_Forensics_Project/score_folder.py
"""
score_folder.py - score a folder of images the MATLAB extractor has measured

For testing against a generator the model has never seen. Extract the folder in
MATLAB first:

    extract_folder('E:\\path\\to\\new_images', 'new.csv')

then score it here, saying what those images actually are:

    python score_folder.py new.csv ai
    python score_folder.py new.csv real
    python score_folder.py new.csv          (predictions only)

None of these images took part in training, so the accuracy reported is a clean
generalisation figure.
"""

import sys
from pathlib import Path

import numpy as np
import joblib

MODEL_PATH = Path(__file__).resolve().parent / "model.joblib"


def main():
    if len(sys.argv) < 2:
        sys.exit(__doc__)

    csv_path = Path(sys.argv[1])
    truth = sys.argv[2].lower() if len(sys.argv) > 2 else None
    if truth not in (None, "ai", "real"):
        sys.exit("Second argument must be 'ai', 'real', or omitted.")
    if not csv_path.exists():
        sys.exit(f"ERROR: cannot find {csv_path}")
    if not MODEL_PATH.exists():
        sys.exit(f"ERROR: cannot find {MODEL_PATH}. Run train_model.py first.")

    X = np.loadtxt(csv_path, delimiter=",")
    if X.ndim == 1:
        X = X.reshape(1, -1)
    # extract_folder.m writes features only; tolerate a trailing label column.
    if X.shape[1] == 231:
        X = X[:, :230]
    if X.shape[1] != 230:
        sys.exit(f"ERROR: expected 230 feature columns, found {X.shape[1]}")

    bundle = joblib.load(MODEL_PATH)
    scores = bundle["model"].predict_proba(X)[:, 1]
    threshold = float(bundle["threshold"])
    pred_ai = scores >= threshold

    names_path = csv_path.with_suffix(".filenames.txt")
    names = (names_path.read_text().splitlines()
             if names_path.exists() and
             len(names_path.read_text().splitlines()) == len(scores)
             else [f"row {i+1}" for i in range(len(scores))])

    print(f"Scored {len(scores)} images from {csv_path.name}")
    print(f"Model threshold: {threshold:.4f}\n")
    print("=" * 58)
    print(f"Predicted AI    : {pred_ai.sum()}  ({pred_ai.mean():.1%})")
    print(f"Predicted real  : {(~pred_ai).sum()}  ({(~pred_ai).mean():.1%})")
    print(f"Score mean/med  : {scores.mean():.3f} / {np.median(scores):.3f}")
    print(f"Score range     : {scores.min():.3f} - {scores.max():.3f}")

    if truth is not None:
        correct = pred_ai if truth == "ai" else ~pred_ai
        print(f"\nGround truth    : all {truth.upper()}")
        print(f"Accuracy        : {correct.mean():.1%}  "
              f"({correct.sum()} / {len(correct)})")
        print("\nNo image here took part in training, so this is a clean "
              "generalisation figure.")

    out = csv_path.with_name(f"scores_{csv_path.stem}.csv")
    with out.open("w") as fh:
        fh.write("filename,score,verdict\n")
        for n, sc in zip(names, scores):
            fh.write(f"{Path(n).name},{sc:.6f},{'AI' if sc >= threshold else 'real'}\n")
    print(f"\nPer-image results: {out.resolve()}")


if __name__ == "__main__":
    main()

File: Generative_Forensics_Project/score_folder.m
function score_folder(folderPath, groundTruth)
%SCORE_FOLDER  Run the trained detector over every image in a folder.
%
%   score_folder('C:\path\to\images')          predictions only
%   score_folder('C:\path\to\images', 'ai')    folder is known to be generated
%   score_folder('C:\path\to\images', 'real')  folder is known to be photographs
%
%   Use this to test the detector against a generator it has never seen. None
%   of these images took part in training, so the accuracy it reports is a
%   clean generalisation number rather than an in-distribution one.
%
%   Writes score_<folder>.csv with one row per image: filename, score, verdict.
%
%   Run train_model.m first.

    MODEL_PATH = 'model.mat';

    if nargin < 2
        groundTruth = '';
    end
    groundTruth = lower(groundTruth);
    if ~isempty(groundTruth) && ~ismember(groundTruth, {'ai', 'real'})
        error('score_folder:badTruth', ...
              'Second argument must be ''ai'', ''real'', or omitted.');
    end
    if ~isfile(MODEL_PATH)
        error('score_folder:noModel', ...
              'Cannot find %s - run train_model.m first.', ...
              fullfile(pwd, MODEL_PATH));
    end
    if ~isfolder(folderPath)
        error('score_folder:noFolder', 'Not a folder: %s', folderPath);
    end

    S = load(MODEL_PATH);

    d = dir(folderPath);
    d = d(~[d.isdir]);
    if isempty(d)
        error('score_folder:empty', 'No files in %s', folderPath);
    end
    names   = {d.name}';
    isImage = ~cellfun(@isempty, regexpi(names, '\.(jpe?g|png)$', 'once'));
    names   = sort(names(isImage));
    if isempty(names)
        error('score_folder:noImages', ...
              'No .jpg/.jpeg/.png images in %s', folderPath);
    end

    fprintf('Scoring %d images from %s\n', numel(names), folderPath);
    fprintf('Model threshold: %.4f\n\n', S.threshold);

    scores  = NaN(numel(names), 1);
    kept    = cell(numel(names), 1);
    nKept   = 0;
    nSmall  = 0;
    nSkip   = 0;

    for i = 1:numel(names)
        f = fullfile(folderPath, names{i});
        try
            info = imfinfo(f);
            if info(1).Width < 256 || info(1).Height < 256
                nSmall = nSmall + 1;
            end

            features    = extractImageFeatures(f, 'crop');
            [~, sc]     = predict(S.model, features);
            nKept       = nKept + 1;
            scores(nKept) = sc(2);
            kept{nKept}   = names{i};
        catch ME
            nSkip = nSkip + 1;
            fprintf(2, 'SKIPPED: %s  (%s)\n', names{i}, ME.message);
        end
        if mod(i, 50) == 0
            fprintf('  %d / %d\n', i, numel(names));
        end
    end

    scores = scores(1:nKept);
    kept   = kept(1:nKept);
    predAI = scores >= S.threshold;

    fprintf('\n==========================================================\n');
    fprintf('Scored          : %d', nKept);
    if nSkip > 0
        fprintf('   (skipped %d)', nSkip);
    end
    fprintf('\n');
    if nSmall > 0
        fprintf('Below 256px     : %d  <- upscaled first, so unreliable\n', nSmall);
    end
    fprintf('Predicted AI    : %d  (%.1f%%)\n', sum(predAI),  100*mean(predAI));
    fprintf('Predicted real  : %d  (%.1f%%)\n', sum(~predAI), 100*mean(~predAI));
    fprintf('Score mean/med  : %.3f / %.3f\n', mean(scores), median(scores));
    fprintf('Score range     : %.3f - %.3f\n', min(scores), max(scores));

    if ~isempty(groundTruth)
        if strcmp(groundTruth, 'ai')
            correct = predAI;
        else
            correct = ~predAI;
        end
        fprintf('\nGround truth    : all %s\n', upper(groundTruth));
        fprintf('Accuracy        : %.1f%%  (%d / %d)\n', ...
                100*mean(correct), sum(correct), numel(correct));
        fprintf(['\nNo image here took part in training, so this is a clean\n' ...
                 'generalisation figure.\n']);
    end

    [~, folderName] = fileparts(strip_trailing_sep(folderPath));
    outPath = sprintf('score_%s.csv', folderName);
    fid = fopen(outPath, 'w');
    if fid == -1
        warning('score_folder:cannotWrite', 'Could not write %s', outPath);
        return;
    end
    cleanup = onCleanup(@() fclose(fid));
    fprintf(fid, 'filename,score,verdict\n');
    for i = 1:nKept
        if scores(i) >= S.threshold
            verdict = 'AI';
        else
            verdict = 'real';
        end
        fprintf(fid, '%s,%.6f,%s\n', kept{i}, scores(i), verdict);
    end
    clear cleanup;

    fprintf('\nPer-image results: %s\n', fullfile(pwd, outPath));
end


function p = strip_trailing_sep(p)
%STRIP_TRAILING_SEP  So fileparts sees the folder name, not an empty tail.
    while ~isempty(p) && (p(end) == '/' || p(end) == '\')
        p(end) = [];
    end
end
